### PR TITLE
Add `util/groupcommit` package in `core` subproject

### PR DESCRIPTION
**Pull request by komamitsu**
## Description

ScalarDB may sometimes need to work with a high latency storage (e.g., multi-region replicated database). In such a case, the throughput of application using ScalarDB would be affected by the high latency transactions. With a group commit mechanism, high latency issue can be hidden with concurrent transactions.

In this PR, we introduce `util/groupcommit` package in `core` subproject which provides a mechanism to support the ScalarDB protocol. The package alone doesn't affect the existing implementations. A following PR will integrate the new package and the existing implementations.

## Related issues and/or PRs

N/A

## Changes made

This PR basically adds the following classes.
image

The `util/groupcommit` package has a concept about keys and they should be described before looking into the classes.

- Keys
  - Child Key : This is passed by a client (e.g., UUID v4 like ScalarDB transaction ID)
  - Parent Key : This is generated by KeyManipulator class which is injected by a client and contains logics to handle Keys
  - Full Key : This consists of Child Key and Parent Key

The main components introduced by this PR are as follows:
- Group
  - A unit of (group) commit
  - Contains slots (see below about slots). The number of slots in a group is decided by the configuration
  - Possible status are as follows:
    - OPEN : Initial status. Accepting new slot reservation as the number of slots isn't fixed yet
    - CLOSED : Not accepting new slot reservations since the number of slots is already fixed. Waiting all the slots are set with values
    - READY : All the slots are set with values. Ready to commit
    - DONE : Group commit is done and all the clients have get the results
  - There are 2 types of Groups
    - NormalGroup
      - Contains multiple slots and commits them at once. Ideally only this Group should be used in terms of performance 
      -  Parent Key is used to identify NormalGroup
    - DelayedGroup
      - Contains only a single slot which is delayed and removed from the NormalGroup
      - Full Key is used to identify DelayedGroup
- Slot
  - A container for a value. Values are supposed to be emitted at once by this group commit mechanism
  - Each Slot has the corresponding Child Key
- GroupManager
  - Manages all the ongoing Groups
  - It internally contains 2 maps
    - NormalGroupMap : Key=Parent-Key, Value=NormalGroup
    - DelayedGroupMap : Key=Full-Key, Value=DelayedGroup
- Worker threads
  - Group-Close Thread : Fixes the number of slot (or close the group) when the Group is timed out
  - Delayed-Slot-Move Thread : Moves delayed Slots from NormalGroupMap to DelayedGroupMap
  - Group-Cleanup Thread : Removes Group entries that are already done from NormalGroupMap or DelayedGroupMap. It's a sort of garbage collector
- GroupCommitter
  - Manages all the above classes
  - Has the following public APIs
    - reserve(Child Key) returns Full Key
      - Allocates a Slot in the current group that is associated with a Full Key
      - The allocated slot status is initially NOT_READY
    - ready(Full Key)
      - Marks the slot associated with the Full Key READY
      - Waits until all the slots in the group are READY
      - Internally calls the `Emittable.emit()` which is injected by the client with either of the following keys once the group gets READY 
        - Parent Key : In most cases, this type of Key is used to emit multiple values in NormalGroup
        - Full Key : This type of  Key is used only when it takes long time to call this API and the delayed slot is separated from the original NormalGroup to a new DelayedGroup to prevent other slots in the Group from being delayed

**Very rough overall architecture of the group commit**
image
Replace `txid` with `Key` when reading the description.

**Lifecycle of Group**
image

## Checklist

> The following is a best-effort checklist. If any items in this checklist are not applicable to this PR or are dependent on other, unmerged PRs, please still mark the checkboxes after you have read and understood each item.

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes.
- [x] Any remaining open issues linked to this PR are documented and up-to-date (Jira, GitHub, etc.).
- [x] Tests (unit, integration, etc.) have been added for the changes.
- [x] My changes generate no new warnings.
- [x] Any dependent changes in other PRs have been merged and published.

## Additional notes (optional)

The current implementation contains the following 3 TODOs. They'll be handled in other PRs.
- Possible performance improvement with tweaking the scope of locks
- Replace the tentative metrics monitoring with DropWizard's one
- Performance improvement of DelayedSlotMoveWorker
  - The worker basically waits for every item, so the performance improvement is required
  - Introducing a priority queue is an option, which was suggested by @brfrn169 

## Release notes

N/A (This PR itself isn't a user facing change).

### CODE REVIEW

File: core/src/main/java/com/scalar/db/util/groupcommit/Group.java
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/groupcommit/Group.java:89]: This shouldn't occur, right? If so, just showing the warning message is okay?

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java:168]: When is this `remove` method intended to be called?

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java:141]: What is the purpose of this `value` argument? Maybe we use it for a transaction ID (a child key) for each transaction within a group?

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java:28]: I think we can put `@ThreadSafe` or `@NotThreadSafe` for each class.

**Review by brfrn169 — COMMENTED** [review]
My first review is complete, and I've left several comments. Please take a look when you have time! I will conduct a second review later on. Thank you for your great work!

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java:90]: Sounds good! I'll update it.

BTW, this will be soon replaced with https://github.com/scalar-labs/scalardb/pull/1614 which uses DropWizard Metrics.

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java:168]: Good question.

Let's say a transaction executes some CRUD operations, starts PREPARE, and fails before the group commit for coordinator state. In this case, a group commit slot is allocated for the transaction and needed to be removed by calling this method not to leave garbage slots.

File: core/src/main/java/com/scalar/db/util/groupcommit/Group.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/Group.java:89]: Good catch! Right. I'll update it to throw AssertionError or IllegalStateException.

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java:141]: In the case of coordinator state commit, `V` is supposed to be `Snapshot` and a passed value will be committed with other snapshots in the same group.

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java]: Fixed in https://github.com/scalar-labs/scalardb/pull/1607/commits/ade2864d14f5902984fb40c553f98380427feef7

File: core/src/main/java/com/scalar/db/util/groupcommit/Group.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/Group.java]: Fixed in https://github.com/scalar-labs/scalardb/pull/1607/commits/8d96fd484d03b1d127d338ac3d128ad8d6232bbc

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java:144]: Updated the comment to make it clearer in https://github.com/scalar-labs/scalardb/pull/1607/commits/fa5e72631b3ca1ef5eee6a47fff13071d9adbd12

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java:172]: Updated the comment in https://github.com/scalar-labs/scalardb/pull/1607/commits/4ff928a8cf45c2e137c75fac93608e73f5b60040

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java:30]: Fixed in https://github.com/scalar-labs/scalardb/pull/1607/commits/e58191815c85f3d27ed4a4c7e8aa8d36549b9976

File: core/src/main/java/com/scalar/db/util/groupcommit/Group.java
Comment (feeblefakie) [inline core/src/main/java/com/scalar/db/util/groupcommit/Group.java:35]: `CLOSED` corresponds to `OPEN` so it's good, but it sounds like the group has already finished processing.
How about `FIXED`?

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupManager.java
Comment (feeblefakie) [inline core/src/main/java/com/scalar/db/util/groupcommit/GroupManager.java:71]: I might miss something, but since we have only one active group at a time, I'm wondering how much it badly affects performance if there are many concurrent transactions.
We could use multiple active groups to mitigate the potential issue, but it might complicate the mechanism.
Any thoughts on this?

**Review by feeblefakie — COMMENTED** [review]
Thank you for the great improvement and very sorry for the late review.
I took a brief look and left some comments and questions. PTAL!
(Let me check this one more time.)

Let me ask additional questions here.
- We discussed before that we might need to access the coordinator table twice in some recovery cases, but how does the current code handle the case?
- If one transaction in a group fails to commit, will the transaction be handled in a `DelayedGroup`?

File: core/src/main/java/com/scalar/db/util/groupcommit/Group.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/Group.java:35]: Thanks for the suggestion! Actually it was first like SIZE-FIXED and I changed it to CLOSED to sound to reject new requests. But as you said, it also sounds things are done. I'll update it to use FIX.

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupManager.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/GroupManager.java:71]: Good point. I also thought and tried multiple current (== active) groups before. But, the performance with multiple active groups didn't have a good performance. I think how fast to fill up a group with slots (and values) is a very important factor to move it to the next state (i.e., ready to commit) in the group commit and using multiple active groups doesn't fit.

I benchmarked this `util/groupcommit` pacakge itself and the average throughput was about 65K TPS without any wait while the throughput with some delays to emulate actual latencies of underlying database had far low throughputs. It means current single active group won't be a bottleneck in usual usages of ScalarDB, I think.

**Comment by komamitsu** [comment]
@feeblefakie Thanks for the questions!

> We discussed before that we might need to access the coordinator table twice in some recovery cases, but how does the current code handle the case?

This PR contains doesn't contain any integration with the coordinator state commit, so let me refer to another PR's change (https://github.com/scalar-labs/scalardb/pull/1523/files#diff-8049e66d11b81cdd5a1dc82c4977b5336741bcee24dd7ddd54c6a6940f09497bR83-R89). I tried the idea to put child-tx-id in a clustering key before in order to make it possible to fetch both group-committed-transactions' ID and separately-committed-transactions' ID. But, there wasn't significant performance improvement with YCSB-F benchmark despite it would require cumbersome table schema migrations. So, current implementation is to access the coordinator table twice.

> If one transaction in a group fails to commit, will the transaction be handled in a DelayedGroup?

No. Slots are moved to DelayedGroup only when it's delayed. If a NormalGroup fails to commit due to network or underlying database issues, all the slots in the group fail. BTW, it's about Commit-State phase. If a transaction fails in Prepare or Validate, it'll be just removed from the NormalGroup without being moved to DelayedGroup.

File: core/src/main/java/com/scalar/db/util/groupcommit/Group.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/Group.java]: Fixed in https://github.com/scalar-labs/scalardb/pull/1607/commits/5000335099a2c15528ef9df89f7ba50e2c4dda82. Used `SIZE_FIXED` to make it clearer.

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupManager.java
Comment (feeblefakie) [reply core/src/main/java/com/scalar/db/util/groupcommit/GroupManager.java:72]: Thank you for the detailed explanation!
It makes sense. Let's leave it as is.

**Comment by feeblefakie** [comment]
@komamitsu 
> I tried the idea to put child-tx-id in a clustering key before in order to make it possible to fetch both group-committed-transactions' ID and separately-committed-transactions' ID. But, there wasn't significant performance improvement with YCSB-F benchmark despite it would require cumbersome table schema migrations. So, current implementation is to access the coordinator table twice.

Thank you for the explanation.
I think how much performance improves depends on how much the workload triggers conflicts,
so it might be worth trying TPC-C to see how it goes since there are not many conflicts in YCSB.

> No. Slots are moved to DelayedGroup only when it's delayed. If a NormalGroup fails to commit due to network or underlying database issues, all the slots in the group fail. BTW, it's about Commit-State phase. If a transaction fails in Prepare or Validate, it'll be just removed from the NormalGroup without being moved to DelayedGroup.

Thank you!
I got it.

File: core/src/main/java/com/scalar/db/util/groupcommit/KeyManipulator.java
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/groupcommit/KeyManipulator.java:43]: Nit: It looks like it's never used.

File: core/src/main/java/com/scalar/db/util/groupcommit/ThrowableRunnable.java
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/groupcommit/ThrowableRunnable.java:4]: We already have `ThrowableRunnable`:
https://github.com/scalar-labs/scalardb/blob/53af316e1badb9460652ef7fcbb54612756edd3b/core/src/main/java/com/scalar/db/util/ThrowableRunnable.java

**Review by feeblefakie — COMMENTED** [review]
Overall, looking good!
Left one question. PTAL!

**Comment by komamitsu** [comment]
> so it might be worth trying TPC-C to see how it goes since there are not many conflicts in YCSB.

@feeblefakie Yeah, good point. I'll try TPC-C with the clustering key version group commit before submitting a PR for the integration of the group commit with the coordinator state commit.

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupManager.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/GroupManager.java]: Fixed it in https://github.com/scalar-labs/scalardb/pull/1607/commits/0007ba39a3d7ccc73ac7064e2fc81193ac0c4fc8

File: core/src/main/java/com/scalar/db/util/groupcommit/ThrowableRunnable.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/ThrowableRunnable.java]: Fixed it in https://github.com/scalar-labs/scalardb/pull/1607/commits/e8faee28e7f1a7f7cdfb2b2d1d36cc8673e34e36

File: core/src/main/java/com/scalar/db/util/groupcommit/BackgroundWorker.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/BackgroundWorker.java:78]: Good point. It shouldn't happen in normal cases. Throwing AssertionError or IllegalStateException is an option. But throwing either of them from background worker doesn't make sense so much and I think just outputting error log is enough.

But, I noticed the error message probably looks unclear thanks to your comment. I updated it in https://github.com/scalar-labs/scalardb/pull/1607/commits/bdf344f560b33f02316c16a1c87e037782eaea5f

**Comment by komamitsu** [comment]
@feeblefakie @brfrn169 Thanks for reviewing this PR!
This is a small FYI, but I renamed `asyncEmit()` to `delegateEmitTaskToWaiter()` in https://github.com/scalar-labs/scalardb/pull/1607/commits/545ad204a0c4ee05931c0a8f983f7ee0225d0217 since the old method name was misleading (actually, the emit task is executed by one of the slots in the group synchronously in `waitUntilEmit()`)

@brfrn169 The performance improvement task of DelayedSlotMoveWorker will be handled in another PR https://github.com/scalar-labs/scalardb/pull/1607/commits/545ad204a0c4ee05931c0a8f983f7ee0225d0217#diff-21491e996128b6ee7175154186a58be20ca714274559ba304eb39c59be8db4faR110-R113

File: core/src/main/java/com/scalar/db/util/groupcommit/Slot.java
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/groupcommit/Slot.java:21]:
```suggestion
  private final CompletableFuture<ThrowableRunnable<Exception>> completableFuture = new CompletableFuture<>();
```

File: core/src/main/java/com/scalar/db/util/groupcommit/Slot.java
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/groupcommit/Slot.java:54]:
```suggestion
      ThrowableRunnable<Exception> taskToEmit = completableFuture.get();
```

File: core/src/main/java/com/scalar/db/util/groupcommit/Slot.java
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/groupcommit/Slot.java:69]:
```suggestion
    } catch (Exception e) {
```

File: core/src/main/java/com/scalar/db/util/groupcommit/Slot.java
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/groupcommit/Slot.java:112]:
```suggestion
  void delegateTaskToWaiter(ThrowableRunnable<Exception> task) {
```

File: core/src/main/java/com/scalar/db/util/groupcommit/NormalGroup.java
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/groupcommit/NormalGroup.java:109]:
```suggestion
    ThrowableRunnable<Exception> taskForEmitterSlot =
```

**Review by brfrn169 — COMMENTED** [review]
Left very minor suggestions. It looks good to me, but let me review this once again later.

File: core/src/main/java/com/scalar/db/util/groupcommit/GroupManager.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/GroupManager.java]: Thanks! Fixed it in https://github.com/scalar-labs/scalardb/pull/1607/commits/865830013b893112149fa88ce705cd60aa465226

File: core/src/main/java/com/scalar/db/util/groupcommit/Slot.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/Slot.java]: Thanks! Fixed it in https://github.com/scalar-labs/scalardb/pull/1607/commits/865830013b893112149fa88ce705cd60aa465226

File: core/src/main/java/com/scalar/db/util/groupcommit/Slot.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/Slot.java]: Thanks! Fixed it in https://github.com/scalar-labs/scalardb/pull/1607/commits/865830013b893112149fa88ce705cd60aa465226

File: core/src/main/java/com/scalar/db/util/groupcommit/NormalGroup.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/groupcommit/NormalGroup.java]: Thanks! Fixed it in https://github.com/scalar-labs/scalardb/pull/1607/commits/865830013b893112149fa88ce705cd60aa465226

File: core/src/main/java/com/scalar/db/util/groupcommit/BackgroundWorker.java
Comment (feeblefakie) [reply core/src/main/java/com/scalar/db/util/groupcommit/BackgroundWorker.java:78]: Looking good!
Yeah, it cannot stop processing even such a case, so having the updated comment makes sense.

**Review by feeblefakie — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**
